### PR TITLE
Snapshoting API

### PR DESCRIPTION
**Pull request by generall**
### All Submissions:

* [ ] Have you followed the guidelines in our Contributing document?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [ ] Does your submission pass tests?
2. [ ] Have you lint your code locally using ``cargo fmt`` command prior to submission?
3. [ ] Have you checked your code using ```cargo clippy``` command?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by agourlay** [comment]
What is exactly the format of the snapshot for a collection?

Is it a directory containing a set of segment archives (tar)?

File: src/actix/api/snapshot_api.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+#[post("/collections/{name}/snapshots")]
+async fn create_snapshot(
+    toc: web::Data<Arc<TableOfContent>>,
+    path: web::Path<String>,
+) -> impl Responder {
+    let collection_name = path.into_inner();
+
+    let timing = Instant::now();
+    let response = do_create_snapshot(&toc.into_inner(), &collection_name).await;
```
Comment (agourlay) [inline src/actix/api/snapshot_api.rs:52]: I wonder if we should keep the user waiting while we are taking the snapshot.
The connection could timeout if the collection is large.

Comment (generall) [reply src/actix/api/snapshot_api.rs:52]: it might be optional, but we definitely need a way to await

File: lib/collection/src/lib.rs [hunk 2]
```diff
@@ -886,6 +898,99 @@ impl Collection {
...
+                "Can't read collection config due to {}\nat {}",
+                err,
+                target_dir.to_str().unwrap()
+            )
+        });
+        let configured_shards = config.params.shard_number.get();
+
+        for shard_id in 0..configured_shards {
+            let shard_path = shard_path(target_dir, shard_id);
+            if shard_path.exists() {
```
Comment (agourlay) [inline lib/collection/src/lib.rs:983]: Why do we need to check for existence here? Is it possible to have a shard missing?

File: openapi/openapi-snapshots.ytt.yaml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    get:
+      tags:
+        - collections
+      summary: List snapshots
+      description: Get list of snapshots for a collection
+      operationId: list_snapshots
+      parameters:
+        - name: collection_name
+          in: path
+          description: Name of the collection to retrieve
```
Comment (agourlay) [inline openapi/openapi-snapshots.ytt.yaml:15]:
```suggestion
          description: Name of the collection
```

File: openapi/openapi-snapshots.ytt.yaml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    post:
+      tags:
+        - collections
+      summary: Create snapshot
+      description: Create new snapshot for a collection
+      operationId: create_snapshot
+      parameters:
+        - name: collection_name
+          in: path
+          description: Name of the collection to retrieve
```
Comment (agourlay) [inline openapi/openapi-snapshots.ytt.yaml:30]:
```suggestion
          description: Name of the collection for which to create a snapshot
```

File: src/main.rs [hunk 5]
```diff
@@ -49,6 +51,17 @@ struct Args {
...
     uri: Option<Uri>,
+
+    /// Force snapshot re-creation
+    /// If provided - existing collections will be replaces with snapshots.
+    /// Default is to not recreate from snapshots.
+    #[clap(short, long, action, default_value_t = false)]
+    force_snapshot: bool,
+
+    /// List of paths to snapshot files.
+    /// Format: <snapshot_file_path>:<collection_name>
```
Comment (agourlay) [inline src/main.rs:62]:
```suggestion
    /// Format: <snapshot_file_path>:<target_collection_name>
```

**Review by agourlay — APPROVED** [review]

**Pull request merged**